benchmarks-mobile/protolite_app/app/src/main/java/io/grpc/grpcbenchmarks/ProtobufBenchmarker.java

```diff
@@ -186,37 +213,54 @@ public void execute() throws JSONException {
         }
     }
 
+    /**
+     * Given an Action, benchmarks it. Calculates how many iterations it takes for a minimum
+     * sample time, and then calculates the number of iterations to run for the target
+     * benchmark time.
+     */
     private static BenchmarkResult benchmark(String name, long dataSize, Action action)
             throws Exception {
+        //TODO: do an actual warmup, much more complicated than originally thought
         for (int i = 0; i < 100; ++i) {
             action.execute();
         }
 
-        System.gc();
+        final AtomicBoolean dead = new AtomicBoolean();
+        Handler handler = new Handler(Looper.getMainLooper());
+        handler.postDelayed(new Runnable() {
+            @Override
+            public void run() {
+                dead.set(true);
+            }
+        }, MIN_SAMPLE_TIME_MS);
 
-        int iterations = 1;
-        long elapsed = timeAction(action, iterations);
-        while (elapsed < MIN_SAMPLE_TIME_MS) {
-            iterations *= 2;
-            elapsed = timeAction(action, iterations);
+        int iterations = 0;
+        for (; !dead.get(); ++iterations) {
+            action.execute();
         }
 
-        iterations = (int) ((TARGET_TIME_MS / (double) elapsed) * iterations);
-        elapsed = timeAction(action, iterations);
-        float mbps = (iterations * dataSize) / (elapsed * 1024 * 1024 / 1000f);
+        iterations = (int) ((TARGET_TIME_MS / MIN_SAMPLE_TIME_MS) * iterations);
+        long elapsed = timeAction(action, iterations);
+        float mbps = (iterations * dataSize) / (elapsed / 1000000000f * 1024 * 1024);
```

Comment:
You really don't want float anywhere unless you have a really good reason.  (regularly old good reasons are not sufficient).

Floats only have 23 bits of accuracy, which is less than the 30 bits you need to represent a billion.  And it's sticky, the loss follows you around, compounding on each calculation.

In general, you should be using longs unless otherwise necessary.